File: src/main/java/com/example/probation/validatiion/validator/NameValidator.java
package com.example.probation.validatiion.validator;

import com.example.probation.service.UsersService;
import com.example.probation.validatiion.annotation.UniqueUsername;
import lombok.RequiredArgsConstructor;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

@RequiredArgsConstructor
public class NameValidator implements ConstraintValidator<UniqueUsername, String> {
    private final UsersService usersService;

    @Override
    public void initialize(UniqueUsername uniqueName) {
        ConstraintValidator.super.initialize(uniqueName);
    }

    @Override
    public boolean isValid(String name, ConstraintValidatorContext context) {
        if (name == null) {
            return false;
        }

        return !usersService.checkUsernameExistence(name);
    }
}

File: src/main/kotlin/com/example/probation/validation/validator/NameValidator.kt
package com.example.probation.validation.validator

import com.example.probation.service.UsersService
import com.example.probation.validation.annotation.UniqueUsername

import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class NameValidator(
    private val usersService: UsersService
    ) : ConstraintValidator<UniqueUsername, String> {

    override fun isValid(name: String?, context: ConstraintValidatorContext) =
        if (name == null) {
            false
        } else !usersService.checkUsernameExistence(name)

    override fun initialize(uniqueUsername: UniqueUsername) {
        super.initialize(uniqueUsername)
    }
}
